src/org/sosy_lab/java_smt/basicimpl/AbstractFormulaManager.java

```diff
@@ -261,11 +265,225 @@ public EnumerationFormulaManager getEnumerationFormulaManager() {
+              throw new IllegalArgumentException();
+            }
+          }
+        } else {
+          // We're inside an r-expression
+          if (c != '\n') {
+            // Append the letter to the token, unless it's a newline
+            token.append(c);
+          } else {
+            // Replace newline with space
+            token.append(' ');
+          }
+          // Handle opening/closing brackets
+          if (c == '(') {
+            level++;
+          }
+          if (c == ')') {
+            if (level == 1) {
+              builder.add(token.toString());
+              token = new StringBuilder();
+            }
+            level--;
+          }
+        }
+      }
+    }
+    if (level != 0) {
+      // Throw an exception if the brackets don't match
+      throw new IllegalArgumentException();
+    }
+    return builder.build();
+  }
+
+  /**
+   * Check if the token is a function or variable declaration.
+   *
+   * <p>Use {@link #tokenize(String)} to turn an SMT-LIB2 script into a string of input tokens.
+   */
+  protected static boolean isDeclarationToken(String token) {
+    return token.matches("\\(\\s*(declare-const|declare-fun).*");
+  }
+
+  /**
+   * Check if the token is a function definition.
+   *
+   * <p>Use {@link #tokenize(String)} to turn an SMT-LIB2 script into a string of input tokens.
+   */
+  protected static boolean isDefinitionToken(String token) {
+    return token.matches("\\(\\s*define-fun.*");
+  }
+
+  /**
+   * Check if the token is an <code>(assert ...)</code>.
+   *
+   * <p>Use {@link #tokenize(String)} to turn an SMT-LIB2 script into a string of input tokens.
+   */
+  protected static boolean isAssertToken(String token) {
+    return token.matches("\\(\\s*assert.*");
+  }
+
+  /**
+   * Check if the token is <code>(set-logic ..)</code>.
+   *
+   * <p>Use {@link #tokenize(String)} to turn an SMT-LIB2 script into a string of input tokens.
+   */
+  protected static boolean isSetLogicToken(String token) {
+    return token.matches("\\(\\s*set-logic.*");
+  }
+
+  /**
+   * Check if the token is <code>(exit ...)</code>.
+   *
+   * <p>Use {@link #tokenize(String)} to turn an SMT-LIB2 script into a string of input tokens.
+   */
+  protected static boolean isExitToken(String token) {
+    return token.matches("\\(\\s*exit.*");
+  }
+
+  /**
+   * Takes a SMT-LIB2 script and cleans it up.
+   *
+   * <p>We remove all comments and put each command on its own line. Declarations and asserts are
+   * kept and everything else is removed. For <code>(set-logic ..)</code> we make sure that it's at
+   * the top of the file before removing it, and for <code>(exit)</code> we make sure that it can
+   * only occur as the last command.
+   */
+  private String sanitize(String formulaStr) {
+    List<String> tokens = tokenize(formulaStr);
+
+    StringBuilder builder = new StringBuilder();
+    int pos = 1; // index of the current token
```

Comment:
Be careful with indices. Developers like to start counting from zero... :smile:  Looks good.

Reply:
Fixed 😉